pthread/pthread.c

```diff
@@ -803,9 +1074,101 @@ void *pthread_getspecific(pthread_key_t key)
 }
 
 
+int pthread_once(pthread_once_t *once_control, void (*init_routine)(void))
+{
+	mutexLock(pthread_common.pthread_key_lock);
+
+	if (*once_control == PTHREAD_ONCE_INIT) {
+		init_routine();
+		*once_control = 0;
+	}
+
+	mutexUnlock(pthread_common.pthread_key_lock);
+	return EOK;
+}
+
+
+int pthread_atfork(void (*prepare)(void), void (*parent)(void), void (*child)(void))
+{
+	int err = EOK;
+	pthread_fork_handlers_t *fork_handlers = (pthread_fork_handlers_t *)malloc(sizeof(pthread_fork_handlers_t));
+	if (fork_handlers == NULL) {
+		err = -ENOMEM;
+	}
+	else {
+		fork_handlers->prepare = prepare;
+		fork_handlers->parent = parent;
+		fork_handlers->child = child;
+		mutexLock(pthread_common.pthread_atfork_lock);
+		LIST_ADD(&pthread_common.pthread_fork_handlers, fork_handlers);
+		mutexUnlock(pthread_common.pthread_atfork_lock);
+	}
+	return err;
+}
+
+
+void _pthread_atfork_prepare(void)
+{
+	/* prepare functions must be called in reverse order */
+	mutexLock(pthread_common.pthread_atfork_lock);
+	pthread_fork_handlers_t *last = pthread_common.pthread_fork_handlers, *curr;
+
+	if (last != NULL) {
+		last = last->prev;
+		curr = last;
+		do {
+			if (curr->prepare != NULL) {
+				curr->prepare();
+			}
+			curr = curr->prev;
+		} while (curr != last);
+	}
+	mutexUnlock(pthread_common.pthread_atfork_lock);
+}
+
+
+/* parent and child functions must be called in order */
+void _pthread_atfork_parent(void)
+{
+	mutexLock(pthread_common.pthread_atfork_lock);
+	pthread_fork_handlers_t *first = pthread_common.pthread_fork_handlers,
+							*curr = pthread_common.pthread_fork_handlers;
+
+	if (first != NULL) {
+		do {
+			if (curr->parent != NULL) {
+				curr->parent();
+			}
+			curr = curr->next;
+		} while (curr != first);
+	}
+	mutexUnlock(pthread_common.pthread_atfork_lock);
+}
+
+
+void _pthread_atfork_child(void)
+{
+	mutexLock(pthread_common.pthread_atfork_lock);
+	pthread_fork_handlers_t *first = pthread_common.pthread_fork_handlers,
+							*curr = pthread_common.pthread_fork_handlers;
+
+	if (first != NULL) {
+		do {
+			if (curr->child != NULL) {
+				curr->child();
+			}
+			curr = curr->next;
+		} while (curr != first);
+	}
+	mutexUnlock(pthread_common.pthread_atfork_lock);
+}
+
+
 void _pthread_init(void)
 {
 	mutexCreate(&pthread_common.pthread_key_lock);
 	mutexCreate(&pthread_common.pthread_list_lock);
+	mutexCreate(&pthread_common.pthread_atfork_lock);
+	mutexCreate(&pthread_common.mutex_cond_init_lock);
```

Comment:
Missing pthread_fork_handlers init (in theory it's ok as is, because it's in .bss, but we do that for pthread_list anyway, so let's do that for pthread_fork_handlers too)